pip/pip-278.md

```diff
@@ -0,0 +1,156 @@
+
+So we need to make  the whole topic compaction service (include Write API & Read API) pluggable to support more customize compaction service implementation.
+
+# Goals
+
+## In Scope
+
+* Abstract topic compaction service interface and support topic compaction service pluggable.
+
+* Migrate the current implementation to a new interface implementation.
+
+* Makes existing tests compatible with new implementations.
+
+## Out of Scope
+
+* For CompactorMetrics, keep the current implementation and don't define related method in topic compaction service interface, in the future will use the `Otel` interface or other mertics API instead.
+
+* For `StrategicTwoPhaseCompactor`, it's out of the scope for regular compaction. It's only used for the load balancer. So it won't change.
+
+
+# High Level Design
+
+To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface, it can provide the capability that the compactor has and provide the read API to read entries from compacted data.
+
+We should combine `CompactedTopicImpl` and `TwoPhaseCompactor` to pulsar implementation of topic compaction service and make behavior with the current implementation consistent.
+
+Class Diagram of core class:
+![pip-278-1.png](images/pip-278-1.png)
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the TopicCompactionService.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can is null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(@NonNUll Position startPosition, int numberOfEntriesToRead);
+
+      /**
+       * Read the last compacted entry from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the compacted last entry, this entry can is null.
+       */
+      CompletableFuture<Entry> readLastCompactedEntry();
+
+      /**
+       * Get the last compacted position from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the last compacted position, this position can is null.
+       */
+      CompletableFuture<Position> getLastCompactedPosition();
```

Comment:
It's confusing that we have both `readLastCompactedEntry` and `getLastCompactedPosition` because we have a `Entry#getPosition` method to get the position of an entry.

Could you explain the difference between:

```java
readLastCompactedEntry().thenApply(Entry::getPosition)
```

and 

```java
getLastCompactedPosition()
```

Reply:
Please ref to: https://github.com/apache/pulsar/pull/20624#discussion_r1236177136